modules/tensor_mechanics/src/materials/ADViscoplasticityStressUpdate.C

```diff
@@ -0,0 +1,290 @@
+  }
+
+  computeStressFinalize(inelastic_strain_increment);
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::initialGuess(const ADReal & effective_trial_stress)
+{
+  return effective_trial_stress;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::maximumPermissibleValue(
+    const ADReal & effective_trial_stress) const
+{
+  return effective_trial_stress * _maximum_gauge_ratio;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::minimumPermissibleValue(
+    const ADReal & effective_trial_stress) const
+{
+  return effective_trial_stress;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::computeResidual(const ADReal & equiv_stress,
+                                                              const ADReal & trial_gauge)
+{
+  const ADReal M = std::abs(_hydro_stress) / trial_gauge;
+  const ADReal dM_dtrial_gauge = -M / trial_gauge;
+
+  const ADReal residual_left = Utility::pow<2>(equiv_stress / trial_gauge);
+  const ADReal dresidual_left_dtrial_gauge = -2.0 * residual_left / trial_gauge;
+
+  ADReal residual = residual_left;
+  _derivative = dresidual_left_dtrial_gauge;
+
+  if (_pore_shape == PoreShapeModel::SPHERICAL)
+  {
+    residual *= 1.0 + _intermediate_porosity / 1.5;
+    _derivative *= 1.0 + _intermediate_porosity / 1.5;
+  }
+
+  if (_model == ViscoplasticityModel::GTN)
+  {
+    residual += 2.0 * _intermediate_porosity * std::cosh(_pore_shape_factor * M) - 1.0 -
+                Utility::pow<2>(_intermediate_porosity);
+    _derivative += 2.0 * _intermediate_porosity * std::sinh(_pore_shape_factor * M) *
+                   _pore_shape_factor * dM_dtrial_gauge;
+  }
+  else
+  {
+    const ADReal h = computeH(_power, M);
+    const ADReal dh_dM = computeH(_power, M, true);
+
+    residual += _intermediate_porosity * (h + _power_factor / h) - 1.0 -
+                _power_factor * Utility::pow<2>(_intermediate_porosity);
+    const ADReal dresidual_dh = _intermediate_porosity * (1.0 - _power_factor / Utility::pow<2>(h));
+    _derivative += dresidual_dh * dh_dM * dM_dtrial_gauge;
+  }
+
+  if (_verbose)
+  {
+    Moose::out << "in computeResidual:\n"
+               << "  position: " << _q_point[_qp] << " hydro_stress: " << _hydro_stress
+               << " equiv_stress: " << equiv_stress << " trial_grage: " << trial_gauge
+               << " M: " << M << std::endl;
+    Moose::out << "  residual: " << residual << "  derivative: " << _derivative << std::endl;
+  }
+
+  return residual;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::computeH(const Real n,
+                                                       const ADReal & M,
+                                                       const bool derivative)
+{
+  const ADReal mod = std::pow(M * _pore_shape_factor, (n + 1.0) / n);
```

Comment:
Can M=0 (division by M below)

Reply:
This is guarded by not running these calculations if there is no equivalent stress and gauge cant be less than equivalent, ie M=equivalent/gauge